File: 11-flag-variations/go/main.go
// Console grid navigator with LaunchDarkly flag variation evaluation.
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"runtime"
	"strings"
	"time"

	"github.com/launchdarkly/go-sdk-common/v3/ldcontext"
	"github.com/launchdarkly/go-sdk-common/v3/ldvalue"
	ld "github.com/launchdarkly/go-server-sdk/v7"
	"github.com/launchdarkly/go-server-sdk/v7/ldcomponents"
	"golang.org/x/term"
)

// LaunchDarkly capability: Multivariate flag evaluation + anonymous contexts
// See: https://launchdarkly.com/docs/sdk/features/flag-types
// See: https://launchdarkly.com/docs/sdk/features/anonymous

const (
	flagAnonOsEmoji = "show-anonymous-host-os-emoji"
	flagCountLabel  = "configure-navigation-count-label"
	flagLuckyNumber = "configure-lucky-number"
	flagMaxMoves    = "configure-max-navigation-moves"
	hostOsAttr      = "hostOs"
	anonymousKey    = "anonymous"

	defaultCountLabel  = "Count"
	defaultLuckyNumber = 0
	defaultMaxMoves    = 100
)

var rows = [3]string{"t", "m", "b"}
var cols = [3]string{"l", "m", "r"}

const bgANSI = "\033[48;5;236m"

var ldClient *ld.LDClient
var hostOS = detectHostOS()

type flagValues struct {
	countLabel  string
	luckyNumber int
	maxMoves    int
	osEmoji     string
}

type position struct {
	row, col int
}

type moveResult struct {
	row, col int
	moved    bool
}

func detectHostOS() string {
	switch runtime.GOOS {
	case "linux":
		return "linux"
	case "windows":
		return "windows"
	case "darwin":
		return "macos"
	default:
		return "other"
	}
}

func osEmojiFor(hostOs string, showOsEmoji bool) string {
	if !showOsEmoji {
		return ""
	}
	switch hostOs {
	case "linux":
		return "🐧"
	case "macos":
		return "🍎"
	case "windows":
		return "🪟"
	default:
		return "😊"
	}
}

func displayName(username, osEmoji string) string {
	if osEmoji == "" {
		return username
	}
	return osEmoji + " " + username
}

func parseMaxMoves(raw ldvalue.Value) int {
	if raw.IsNull() {
		return defaultMaxMoves
	}
	if raw.Type() == ldvalue.ObjectType {
		if v := raw.GetByKey("maxMoves"); v.IsNumber() {
			return int(v.IntValue())
		}
	}
	if raw.IsString() {
		var parsed map[string]interface{}
		if err := json.Unmarshal([]byte(raw.StringValue()), &parsed); err == nil {
			if v, ok := parsed["maxMoves"].(float64); ok {
				return int(v)
			}
		}
	}
	return defaultMaxMoves
}

func buildFlagResponse(countLabel string, luckyNumber, maxMoves int, osEmoji string) flagValues {
	if countLabel == "" {
		countLabel = defaultCountLabel
	}
	return flagValues{
		countLabel:  countLabel,
		luckyNumber: luckyNumber,
		maxMoves:    maxMoves,
		osEmoji:     osEmoji,
	}
}

func defaults() flagValues {
	return buildFlagResponse(defaultCountLabel, defaultLuckyNumber, defaultMaxMoves, "")
}

func initLaunchDarkly() {
	sdkKey := os.Getenv("LD_SDK_KEY")
	if sdkKey == "" {
		fmt.Fprintln(os.Stderr, "Warning: LD_SDK_KEY not set — flags default to off.")
		return
	}
	config := ld.Config{
		Events: ldcomponents.SendEvents().PrivateAttributes(hostOsAttr),
	}
	client, err := ld.MakeCustomClient(sdkKey, config, 5*time.Second)
	if err != nil {
		fmt.Fprintln(os.Stderr, "Warning: LaunchDarkly SDK did not initialize — flags default to off.")
		return
	}
	ldClient = client
}

func evaluateFlags(username string) flagValues {
	if ldClient == nil {
		return defaults()
	}
	anonContext := ldcontext.NewBuilder(anonymousKey).
		Anonymous(true).
		SetString(hostOsAttr, hostOS).
		Private(hostOsAttr).
		Build()
	userContext := ldcontext.NewBuilder(username).Build()

	showEmoji, _ := ldClient.BoolVariation(flagAnonOsEmoji, anonContext, false)
	countLabel, _ := ldClient.StringVariation(flagCountLabel, userContext, defaultCountLabel)
	luckyNumber, _ := ldClient.IntVariation(flagLuckyNumber, userContext, defaultLuckyNumber)
	maxMovesRaw, _ := ldClient.JSONVariation(
		flagMaxMoves,
		userContext,
		ldvalue.ObjectBuild().Set("maxMoves", ldvalue.Int(defaultMaxMoves)).Build(),
	)

	return buildFlagResponse(
		countLabel,
		luckyNumber,
		parseMaxMoves(maxMovesRaw),
		osEmojiFor(hostOS, showEmoji),
	)
}

func formatPos(row, col int) string {
	return rows[row] + "/" + cols[col]
}

func tryMove(row, col, dr, dc int) moveResult {
	nr := clamp(row+dr, 0, 2)
	nc := clamp(col+dc, 0, 2)
	if nr == row && nc == col {
		return moveResult{row, col, false}
	}
	return moveResult{nr, nc, true}
}

func clamp(v, lo, hi int) int {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

func readUsername(reader *bufio.Reader) (string, error) {
	fmt.Println("Login\n")
	for {
		fmt.Print("Username: ")
		line, err := reader.ReadString('\n')
		if err != nil {
			return "", err
		}
		name := strings.TrimSpace(line)
		if name != "" {
			return name, nil
		}
		fmt.Println("Username is required.")
	}
}

func drawCell(selected bool, line int) string {
	if selected {
		switch line {
		case 0:
			return "┏━━━┓"
		case 1:
			return "┃ X ┃"
		default:
			return "┗━━━┛"
		}
	}
	switch line {
	case 0:
		return "┌───┐"
	case 1:
		return "│   │"
	default:
		return "└───┘"
	}
}

func writeLine(out *strings.Builder, line string) {
	out.WriteString(line)
	out.WriteString("\r\n")
}

func render(username string, row, col int, previous *position, moveCount int, flags flagValues) {
	var out strings.Builder
	out.WriteString(bgANSI)
	out.WriteString("\033[H\033[2J")

	prevText := "—"
	if previous != nil {
		prevText = formatPos(previous.row, previous.col)
	}

	writeLine(&out, fmt.Sprintf("Name: %s", displayName(username, flags.osEmoji)))
	writeLine(&out, fmt.Sprintf("Current position: %s", formatPos(row, col)))
	writeLine(&out, fmt.Sprintf("Previous position: %s", prevText))
	writeLine(&out, fmt.Sprintf("%s: %d", flags.countLabel, moveCount))
	writeLine(&out, fmt.Sprintf("Lucky Number is: %d", flags.luckyNumber))
	writeLine(&out, "")
	writeLine(&out, "Use arrow keys or WASD to move (L to logout, Q to quit).")
	writeLine(&out, "")

	for r := 0; r < 3; r++ {
		top := make([]string, 3)
		mid := make([]string, 3)
		bot := make([]string, 3)
		for c := 0; c < 3; c++ {
			selected := r == row && c == col
			top[c] = drawCell(selected, 0)
			mid[c] = drawCell(selected, 1)
			bot[c] = drawCell(selected, 2)
		}
		writeLine(&out, strings.Join(top, " "))
		writeLine(&out, strings.Join(mid, " "))
		writeLine(&out, strings.Join(bot, " "))
	}

	fmt.Print(out.String())
}

type sessionAction int

const (
	actionQuit sessionAction = iota
	actionLogout
)

func readKeyWithTimeout(reader *bufio.Reader, timeout time.Duration) (dr, dc int, action sessionAction, endSession, ok, timedOut bool) {
	if err := os.Stdin.SetReadDeadline(time.Now().Add(timeout)); err != nil {
		return 0, 0, actionQuit, false, false, true
	}
	defer os.Stdin.SetReadDeadline(time.Time{})

	b, err := reader.ReadByte()
	if err != nil {
		if ne, ok := err.(interface{ Timeout() bool }); ok && ne.Timeout() {
			return 0, 0, actionQuit, false, false, true
		}
		return 0, 0, actionQuit, true, false, false
	}

	switch b {
	case 3, 'q', 'Q':
		return 0, 0, actionQuit, true, true, false
	case 'l', 'L':
		return 0, 0, actionLogout, true, true, false
	case 'w', 'W':
		return -1, 0, 0, false, true, false
	case 's', 'S':
		return 1, 0, 0, false, true, false
	case 'a', 'A':
		return 0, -1, 0, false, true, false
	case 'd', 'D':
		return 0, 1, 0, false, true, false
	case 27:
		b2, err := reader.ReadByte()
		if err != nil || b2 != '[' {
			return 0, 0, 0, false, false, false
		}
		b3, err := reader.ReadByte()
		if err != nil {
			return 0, 0, 0, false, false, false
		}
		switch b3 {
		case 'A':
			return -1, 0, 0, false, true, false
		case 'B':
			return 1, 0, 0, false, true, false
		case 'C':
			return 0, 1, 0, false, true, false
		case 'D':
			return 0, -1, 0, false, true, false
		}
	}
	return 0, 0, 0, false, false, false
}

func runGrid(username string, reader *bufio.Reader) sessionAction {
	row, col := 1, 1
	var previous *position
	moveCount := 0

	for {
		flags := evaluateFlags(username)
		render(username, row, col, previous, moveCount, flags)

		dr, dc, action, endSession, ok, timedOut := readKeyWithTimeout(reader, 500*time.Millisecond)
		if timedOut {
			continue
		}
		if endSession {
			return action
		}
		if !ok {
			continue
		}

		if moveCount >= flags.maxMoves {
			continue
		}

		result := tryMove(row, col, dr, dc)
		if result.moved {
			prev := position{row, col}
			previous = &prev
			row, col = result.row, result.col
			moveCount++
		}
	}
}

func main() {
	initLaunchDarkly()
	defer func() {
		if ldClient != nil {
			ldClient.Close()
		}
	}()

	reader := bufio.NewReader(os.Stdin)
	fd := int(os.Stdin.Fd())

	for {
		username, err := readUsername(reader)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}

		oldState, err := term.MakeRaw(fd)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}

		action := runGrid(username, reader)
		term.Restore(fd, oldState)

		if action == actionQuit {
			break
		}
	}
}

File: 11-flag-variations/rust/src/main.rs
//! Console grid navigator with LaunchDarkly flag variation evaluation.

use crossterm::{
    cursor::{Hide, MoveTo, Show},
    event::{self, Event, KeyCode, KeyEvent, KeyModifiers},
    execute, queue,
    style::Print,
    terminal::{self, ClearType},
};
use launchdarkly_server_sdk::{AttributeValue, Client, ConfigBuilder, ContextBuilder};
use std::io::{self, Write};
use std::sync::Arc;
use std::time::Duration;

// LaunchDarkly capability: Multivariate flag evaluation + anonymous contexts
// See: https://launchdarkly.com/docs/sdk/features/flag-types
// See: https://launchdarkly.com/docs/sdk/features/anonymous

const FLAG_ANON_OS_EMOJI: &str = "show-anonymous-host-os-emoji";
const FLAG_COUNT_LABEL: &str = "configure-navigation-count-label";
const FLAG_LUCKY_NUMBER: &str = "configure-lucky-number";
const FLAG_MAX_MOVES: &str = "configure-max-navigation-moves";
const HOST_OS_ATTR: &str = "hostOs";
const ANONYMOUS_KEY: &str = "anonymous";

const DEFAULT_COUNT_LABEL: &str = "Count";
const DEFAULT_LUCKY_NUMBER: i64 = 0;
const DEFAULT_MAX_MOVES: i32 = 100;

const ROWS: [&str; 3] = ["t", "m", "b"];
const COLS: [&str; 3] = ["l", "m", "r"];
const BG: &str = "\x1b[48;5;236m";

struct Position {
    row: i32,
    col: i32,
}

struct MoveResult {
    row: i32,
    col: i32,
    moved: bool,
}

#[derive(PartialEq, Eq)]
enum SessionAction {
    Quit,
    Logout,
}

struct FlagValues {
    count_label: String,
    lucky_number: i32,
    max_moves: i32,
    os_emoji: String,
}

struct App {
    client: Option<Arc<Client>>,
    host_os: String,
}

fn detect_host_os() -> String {
    match std::env::consts::OS {
        "linux" => "linux",
        "windows" => "windows",
        "macos" => "macos",
        _ => "other",
    }
    .to_string()
}

fn os_emoji_for(host_os: &str, show_os_emoji: bool) -> String {
    if !show_os_emoji {
        return String::new();
    }
    match host_os {
        "linux" => "🐧".to_string(),
        "macos" => "🍎".to_string(),
        "windows" => "🪟".to_string(),
        _ => "😊".to_string(),
    }
}

fn display_name(username: &str, os_emoji: &str) -> String {
    if os_emoji.is_empty() {
        username.to_string()
    } else {
        format!("{os_emoji} {username}")
    }
}

fn parse_max_moves(raw: &serde_json::Value) -> i32 {
    raw.get("maxMoves")
        .and_then(|v| v.as_i64())
        .map(|n| n as i32)
        .unwrap_or(DEFAULT_MAX_MOVES)
}

fn build_flag_response(count_label: String, lucky_number: i32, max_moves: i32, os_emoji: String) -> FlagValues {
    FlagValues {
        count_label: if count_label.is_empty() {
            DEFAULT_COUNT_LABEL.to_string()
        } else {
            count_label
        },
        lucky_number,
        max_moves,
        os_emoji,
    }
}

fn defaults() -> FlagValues {
    build_flag_response(
        DEFAULT_COUNT_LABEL.to_string(),
        DEFAULT_LUCKY_NUMBER as i32,
        DEFAULT_MAX_MOVES,
        String::new(),
    )
}

impl App {
    fn new() -> Self {
        let sdk_key = std::env::var("LD_SDK_KEY").ok();
        let client = sdk_key.and_then(|key| {
            let config = ConfigBuilder::new(&key).build().ok()?;
            let client = Client::build(config).ok()?;
            client.start_with_runtime().ok()?;

            let deadline = std::time::Instant::now() + Duration::from_secs(5);
            while !client.initialized() && std::time::Instant::now() < deadline {
                std::thread::sleep(Duration::from_millis(50));
            }
            if client.initialized() {
                Some(Arc::new(client))
            } else {
                eprintln!("Warning: LaunchDarkly SDK did not initialize — flags default to off.");
                None
            }
        });
        if std::env::var("LD_SDK_KEY").is_err() {
            eprintln!("Warning: LD_SDK_KEY not set — flags default to off.");
        }
        Self {
            client,
            host_os: detect_host_os(),
        }
    }

    fn evaluate_flags(&self, username: &str) -> FlagValues {
        let Some(client) = &self.client else {
            return defaults();
        };

        let anon_context = ContextBuilder::new(ANONYMOUS_KEY)
            .anonymous(true)
            .set_value(
                HOST_OS_ATTR,
                AttributeValue::String(self.host_os.clone().into()),
            )
            .add_private_attribute(HOST_OS_ATTR)
            .build()
            .unwrap_or_else(|_| ContextBuilder::new(ANONYMOUS_KEY).build().unwrap());

        let user_context = ContextBuilder::new(username)
            .build()
            .unwrap_or_else(|_| ContextBuilder::new("anonymous").build().unwrap());

        let show_emoji = client.bool_variation(&anon_context, FLAG_ANON_OS_EMOJI, false);
        let count_label = client.str_variation(
            &user_context,
            FLAG_COUNT_LABEL,
            DEFAULT_COUNT_LABEL.to_string(),
        );
        let lucky_number = client.int_variation(&user_context, FLAG_LUCKY_NUMBER, DEFAULT_LUCKY_NUMBER);
        let max_moves_raw = client.json_variation(
            &user_context,
            FLAG_MAX_MOVES,
            serde_json::json!({ "maxMoves": DEFAULT_MAX_MOVES }),
        );

        build_flag_response(
            count_label,
            lucky_number as i32,
            parse_max_moves(&max_moves_raw),
            os_emoji_for(&self.host_os, show_emoji),
        )
    }
}

impl Drop for App {
    fn drop(&mut self) {
        if let Some(client) = self.client.take() {
            if let Ok(client) = Arc::try_unwrap(client) {
                client.close();
            }
        }
    }
}

fn format_pos(row: i32, col: i32) -> String {
    format!("{}/{}", ROWS[row as usize], COLS[col as usize])
}

fn try_move(row: i32, col: i32, dr: i32, dc: i32) -> MoveResult {
    let nr = (row + dr).clamp(0, 2);
    let nc = (col + dc).clamp(0, 2);
    if nr == row && nc == col {
        MoveResult { row, col, moved: false }
    } else {
        MoveResult {
            row: nr,
            col: nc,
            moved: true,
        }
    }
}

fn read_username() -> io::Result<String> {
    println!("Login\n");
    loop {
        print!("Username: ");
        io::stdout().flush()?;
        let mut line = String::new();
        io::stdin().read_line(&mut line)?;
        let name = line.trim().to_string();
        if !name.is_empty() {
            return Ok(name);
        }
        println!("Username is required.");
    }
}

fn cell_line(selected: bool, line: usize) -> &'static str {
    if selected {
        match line {
            0 => "┏━━━┓",
            1 => "┃ X ┃",
            _ => "┗━━━┛",
        }
    } else {
        match line {
            0 => "┌───┐",
            1 => "│   │",
            _ => "└───┘",
        }
    }
}

fn print_line(out: &mut impl Write, y: u16, text: &str) -> io::Result<()> {
    queue!(out, MoveTo(0, y), Print(text))?;
    Ok(())
}

fn render(
    out: &mut impl Write,
    username: &str,
    row: i32,
    col: i32,
    previous: Option<&Position>,
    move_count: i32,
    flags: &FlagValues,
) -> io::Result<()> {
    execute!(out, MoveTo(0, 0), terminal::Clear(ClearType::All))?;
    queue!(out, Print(BG))?;

    let prev_text = previous
        .map(|p| format_pos(p.row, p.col))
        .unwrap_or_else(|| "—".to_string());

    let mut y = 0u16;
    print_line(
        out,
        y,
        &format!("Name: {}", display_name(username, &flags.os_emoji)),
    )?;
    y += 1;
    print_line(out, y, &format!("Current position: {}", format_pos(row, col)))?;
    y += 1;
    print_line(out, y, &format!("Previous position: {prev_text}"))?;
    y += 1;
    print_line(out, y, &format!("{}: {move_count}", flags.count_label))?;
    y += 1;
    print_line(out, y, &format!("Lucky Number is: {}", flags.lucky_number))?;
    y += 2;
    print_line(out, y, "Use arrow keys or WASD to move (L to logout, Q to quit).")?;
    y += 2;

    for r in 0..3 {
        let top = (0..3)
            .map(|c| cell_line(r == row && c == col, 0))
            .collect::<Vec<_>>()
            .join(" ");
        let mid = (0..3)
            .map(|c| cell_line(r == row && c == col, 1))
            .collect::<Vec<_>>()
            .join(" ");
        let bot = (0..3)
            .map(|c| cell_line(r == row && c == col, 2))
            .collect::<Vec<_>>()
            .join(" ");
        print_line(out, y, &top)?;
        y += 1;
        print_line(out, y, &mid)?;
        y += 1;
        print_line(out, y, &bot)?;
        y += 1;
    }

    out.flush()?;
    Ok(())
}

fn run_grid(out: &mut impl Write, app: &App, username: &str) -> io::Result<SessionAction> {
    let mut row = 1;
    let mut col = 1;
    let mut previous: Option<Position> = None;
    let mut move_count = 0;

    loop {
        let flags = app.evaluate_flags(username);
        render(out, username, row, col, previous.as_ref(), move_count, &flags)?;

        if !event::poll(Duration::from_millis(500))? {
            continue;
        }
        if let Event::Key(KeyEvent { code, modifiers, .. }) = event::read()? {
            if modifiers.contains(KeyModifiers::CONTROL) {
                return Ok(SessionAction::Quit);
            }
            let movement = match code {
                KeyCode::Char('q') | KeyCode::Char('Q') => return Ok(SessionAction::Quit),
                KeyCode::Char('l') | KeyCode::Char('L') => return Ok(SessionAction::Logout),
                KeyCode::Up | KeyCode::Char('w') | KeyCode::Char('W') => Some((-1, 0)),
                KeyCode::Down | KeyCode::Char('s') | KeyCode::Char('S') => Some((1, 0)),
                KeyCode::Left | KeyCode::Char('a') | KeyCode::Char('A') => Some((0, -1)),
                KeyCode::Right | KeyCode::Char('d') | KeyCode::Char('D') => Some((0, 1)),
                _ => None,
            };
            let Some((dr, dc)) = movement else {
                continue;
            };
            if move_count >= flags.max_moves {
                continue;
            }
            let result = try_move(row, col, dr, dc);
            if result.moved {
                previous = Some(Position { row, col });
                row = result.row;
                col = result.col;
                move_count += 1;
            }
        }
    }
}

fn main() -> io::Result<()> {
    let app = App::new();
    let mut stdout = io::stdout();

    loop {
        let username = read_username()?;
        terminal::enable_raw_mode()?;
        execute!(stdout, Hide)?;

        let action = run_grid(&mut stdout, &app, &username)?;

        execute!(stdout, Show)?;
        terminal::disable_raw_mode()?;

        if action == SessionAction::Quit {
            break;
        }
    }

    Ok(())
}
